### PR TITLE
[common] Fix #1494 potential for buffer overrun

### LINKED ISSUE #1494: [common] Potential buffer overrun protection failure in km_kbp_context_to_utfX() functions
**Description**
`km_kbp_context_to_utfX()` can potentially overrun the user supplied buffer by 1-3 bytes if the last USV would be encoded as a multi unit code point and the end of the buffer lies between code point boundaries. In each case the function will correctly return `KM_KBP_STATUS_INSUFFICENT_BUFFER` but only after writing past the end of the buffer.

**To Reproduce**
```
km_kbp_context_item test[] = {
    {KM_KBP_CT_CHAR, km_kbp_usv(U'🐛')},   // <U+D83D U+DC1B>
    KM_KBP_CONTEXT_ITEM_END
};

km_kbp_cp buf[2] = {0,};
size_t    buf_sz = 1;
assert(km_kbp_context_to_utf16(test, buf, &buf_sz) ==  KM_KBP_STATUS_INSUFFICENT_BUFFER);
assert(buf_sz == 1);
```

Will fail on the 2nd assert and the 2nd surrogate will have been output to `buf[1]`, when it should not have.

**Expected behavior**
buf_sz should equal 1 after call and only `buf[0]` should have been written to.

---

**Keyman Desktop, Keyman for macOS, Keyman for Linux (please complete the following information):**
 - OS: Any Linux
 - Keyman version: 11 beta
  
---
**Additional context**
This bug will not trigger when using the buffer size count returned by `km_kbp_context_to_utf16(test, 0, &buf_sz)`. When used with correctly sized buffer there is no problem.

### PR DESCRIPTION
**Pull request by tim-eves**
iterators, now deduce and track remaining buffer space from `==` or `!=` comparision operators.  They default to infinite (> 4) when no comparisions are called, making them still useful for scan until terminator uses. This is supplied to every call of get and put.

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/138833:id/kvk/index.html)

File: common/engine/keyboardprocessor/src/utfcodec.hpp [hunk 1]
```diff
@@ -126,31 +128,30 @@ struct _utf_codec<8>
     inline
     static void put(codeunit_t * cp, const uchar_t usv, int8_t & l) throw()
     {
-        if (usv < 0x80)     {l = 1; cp[0] = usv; return; }
-        if (usv < 0x0800)   {l = 2; cp[0] = 0xC0 + (usv >> 6);  cp[1] = 0x80 + (usv & 0x3F); return; }
-        if (usv < 0x10000)  {l = 3; cp[0] = 0xE0 + (usv >> 12); cp[1] = 0x80 + ((usv >> 6) & 0x3F);  cp[2] = 0x80 + (usv & 0x3F); return; }
-        else                {l = 4; cp[0] = 0xF0 + (usv >> 18); cp[1] = 0x80 + ((usv >> 12) & 0x3F); cp[2] = 0x80 + ((usv >> 6) & 0x3F); cp[3] = 0x80 + (usv & 0x3F); return; }
+        --l; if (usv < 0x80)     {l = 1; cp[0] = usv; return; };
```
Comment (mcdurdin) [inline common/engine/keyboardprocessor/src/utfcodec.hpp]: What's with the additional `;` at EOL?

**Comment by mcdurdin** [comment]
Just spotted: this needs to target `beta`, not `master`.

**Comment by tim-eves** [comment]
PR should be based off of beta branch not master

**Pull request closed without merging**